### PR TITLE
Trigger parameter issue

### PR DESCRIPTION
**Pull request by steven0711dong**
close https://github.com/apache/openwhisk-cli/issues/290, https://github.com/apache/openwhisk-cli/issues/293

### CODE REVIEW

File: tests/src/test/scala/system/basic/WskCliBasicTests.scala [hunk 1]
```diff
@@ -451,7 +451,7 @@ class WskCliBasicTests extends TestHelpers with WskTestHelpers {
     getJSONFromResponse(trigger.stdout, true).fields("parameters") shouldBe JsArray(
       JsObject("key" -> JsString("a"), "value" -> JsString("A")))
     getJSONFromResponse(trigger.stdout, true).fields("publish") shouldBe false.toJson
-    getJSONFromResponse(trigger.stdout, true).fields("version") shouldBe "0.0.2".toJson
+    getJSONFromResponse(trigger.stdout, true).fields("version") shouldBe "0.0.1".toJson
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskCliBasicTests.scala]: this is semantic changing - it suggests there was a call to trigger update somewhere before that is no longer occurring.

in fact on line 438 and 439 you see there was a `create` and then an `update`. So I think this change is not correct and should be reverted.

File: commands/trigger.go [hunk 2]
```diff
@@ -594,10 +579,42 @@ func (t *Trigger) Update(Client *whisk.Client, args []string) error {
 			}
 		}
 
+		//if there is no feed attached to this trigger
+		if len(fullFeedName) < 1 {
+			//but user indicate feed parameter change, we issue error message.
+			if len(Flags.trigger.feedParam) > 0 {
+				whisk.Debug(whisk.DbgError, "trigger %+v is found but it does not contain a feed. \n", trigger)
+				errStr := wski18n.T("triggerDoesNotContainFeed", map[string]interface{}{"name": qualifiedName.GetEntityName()})
```
Comment (rabbah) [inline commands/trigger.go]: Please look at using a constant in https://github.com/apache/openwhisk-cli/blob/master/commands/messages.go instead of a string `triggerDoesNotContainFeed` which requires a matching entry in the localization file.

Comment (steven0711dong) [reply commands/trigger.go]: `triggerDoesNotContainFeed` represents a very specific message that tells users they've run into a specific situation that will results to error. This happens when they tries to update feed params when the trigger contains no feed. I cannot find any constant in messages.go file that represents this. There is only one constant that's relevant, which is `FEED_CONFIGURATION_FAILURE`, but this is too broad and general error message.

File: commands/commands.go [hunk 3]
```diff
@@ -191,7 +192,17 @@ func parseArgs(args []string) ([]string, []string, []string, error) {
...
 				whiskErr = whisk.MakeWskError(errors.New(errMsg), whisk.EXIT_CODE_ERR_GENERAL, whisk.DISPLAY_MSG,
 					whisk.DISPLAY_USAGE)
-				return nil, nil, nil, whiskErr
+				return nil, nil, nil, nil, whiskErr
+			}
+		} else if args[i] == "-fp" || args[i] == "--feed-param" {
+			feedParamArgs, args, whiskErr = getKeyValueArgs(args, i, feedParamArgs)
+			if whiskErr != nil {
+				whisk.Debug(whisk.DbgError, "getKeyValueArgs(%#v, %d) failed: %s\n", args, i, whiskErr)
+				errMsg := wski18n.T("The parameter arguments are invalid: {{.err}}",
```
[the remark this answers is not in the document]
Comment (steven0711dong) [reply commands/commands.go]: I realize that it is sufficient to rely on error message produced by getKeyValueArgs. So I removed the unnecessary error string.

File: commands/trigger.go (hunk 2 above)
Comment (steven0711dong) [reply commands/trigger.go]: You are correct. I do not need to add another string in the resource file, so I've changed it.

File: tests/src/test/scala/system/basic/WskCliBasicTests.scala (hunk 1 above)
Comment (steven0711dong) [reply tests/src/test/scala/system/basic/WskCliBasicTests.scala]: Nice catch!

File: commands/trigger.go [hunk 4]
```diff
@@ -436,47 +440,118 @@ func (t *Trigger) Create(Client *whisk.Client, args []string) error {
...
 
+	//simplestTrigger indicates user are creating a trigger without any feed or parameters
+	simplestTrigger := len(Flags.trigger.feedParam) == 0 && len(Flags.common.param) == 0 && len(Flags.trigger.triggerParam) == 0
+
+	//if users are 1. creating a trigger without any feed or parameters
+	//             2. creating a trigger using --param flag
+	//then we use the old way to create the trigger.
+	if len(Flags.common.param) > 0 || simplestTrigger {
+		if len(Flags.trigger.feedParam) > 0 || len(Flags.trigger.triggerParam) > 0 {
+			whisk.Debug(whisk.DbgError, "User tries to mix use of --param with --feed-param and --trigger-param")
```
Comment (dubee) [inline commands/trigger.go]: Is there a use case for the user trying to use trigger parameters and feed parameters together?

File: commands/trigger.go [hunk 5]
```diff
@@ -436,47 +440,118 @@ func (t *Trigger) Create(Client *whisk.Client, args []string) error {
...
 	// if a feed is specified, add feed annotation the annotations declared on the command line
 	// TODO: add test to ensure that generated annotation has precedence
-	if feedName != nil {
-		annotationArray = append(annotationArray, getFormattedJSON("feed", feedName.GetFullQualifiedName()))
+	if feedQualifiedName != nil {
+		annotationArray = append(annotationArray, getFormattedJSON("feed", feedQualifiedName.GetFullQualifiedName()))
 	}
 	annotations := getParameters(annotationArray, true, true)
 
+	//simplestTrigger indicates user are creating a trigger without any feed or parameters
```
Comment (dubee) [inline commands/trigger.go]: Does duplicate code exist in create and update? If so it'd be better to create functions out of the duplicate code to reduce complexity.

File: commands/commands.go [hunk 6]
```diff
@@ -181,17 +184,34 @@ func parseArgs(args []string) ([]string, []string, []string, error) {
 					map[string]interface{}{"err": whiskErr})
 				whiskErr = whisk.MakeWskError(errors.New(errMsg), whisk.EXIT_CODE_ERR_GENERAL, whisk.DISPLAY_MSG,
 					whisk.DISPLAY_USAGE)
-				return nil, nil, nil, whiskErr
+				return nil, nil, nil, nil, nil, whiskErr
 			}
 		} else if args[i] == "-a" || args[i] == "--annotation" {
+			fmt.Println("4")
```
[the remark this answers is not in the document]
Comment (steven0711dong) [reply commands/commands.go]: oops..

File: commands/trigger.go (hunk 4 above)
Comment (steven0711dong) [reply commands/trigger.go]: Yeah, if users want to create a trigger using the new way. They will create trigger parameters using --trigger-param flag and create feed parameters using --feed-param flag. They just cannot combine --param with --feed-param or --trigger-param. The meaning is ambiguous. So we issue an error here.

File: commands/trigger.go (hunk 5 above)
Comment (steven0711dong) [reply commands/trigger.go]: I created two functions so that both create and update could use. Some other variables such as params, feed params do exist in both functions but are created differently and the creation depends on users' intention on how to issue trigger command(old way and new way). So it might make things more complicated to create functions out of duplicate code. 

I have to admit that after adding the new flags, both functions become bigger but I think it might be a good idea to separate execution path for the old way of creating trigger from the new way so they do not get lumped together.

File: commands/commands.go (hunk 6 above)
Comment (dubee) [inline commands/commands.go]: Remove print statement.

File: commands/commands.go [hunk 7]
```diff
@@ -19,6 +19,7 @@ package commands
 
 import (
 	"errors"
+	"fmt"
```
Comment (dubee) [inline commands/commands.go]: Remove `fmt`.

File: commands/trigger.go [hunk 8]
```diff
@@ -436,47 +442,108 @@ func (t *Trigger) Create(Client *whisk.Client, args []string) error {
...
+		return nil
+	}
+
+	//if execution reached this line, it means that users intend to create or update trigger using the new way
+	//by using --feed-param or --trigger-param flag
+	if feedQualifiedName == nil && len(Flags.trigger.feedParam) > 0 {
+		//trigger without feed names cannot have a feed param
+		errMsg := fmt.Sprintf("Incorrect usage. trigger without a feed cannot have feed parameters. \n")
+		fmt.Fprintf(os.Stderr, "%s%s", color.RedString("error: "), errors.New(errMsg))
+		return nil
```
Comment (dubee) [inline commands/trigger.go]: Hmm, errors are usually return back to the system instead of `nil`.

File: commands/trigger.go [hunk 9]
```diff
@@ -597,7 +679,75 @@ func (t *Trigger) Update(Client *whisk.Client, args []string) error {
...
+//then we use the old way to create the trigger.
+func userIndicatesToUseOldTriggerCommand() bool {
+	return len(Flags.common.param) > 0 || (len(Flags.trigger.feedParam) == 0 && len(Flags.common.param) == 0 && len(Flags.trigger.triggerParam) == 0)
+}
+
+func userIssuedNewTriggerCmd() bool {
+	if len(Flags.trigger.feedParam) > 0 || len(Flags.trigger.triggerParam) > 0 {
+		whisk.Debug(whisk.DbgError, "User tries to mix use of --param with --feed-param and --trigger-param")
+		errMsg := fmt.Sprintf("Incorrect usage. Cannot combine --feed-param or --trigger-param flag with --param flag\n")
+		fmt.Fprintf(os.Stderr, "%s%s", color.RedString("error: "), errors.New(errMsg))
```
Comment (dubee) [inline commands/trigger.go]: Errors are usually returned to the system, which halts further execution.

File: commands/trigger.go [hunk 10]
```diff
@@ -436,47 +442,108 @@ func (t *Trigger) Create(Client *whisk.Client, args []string) error {
...
 	// if a feed is specified, add feed annotation the annotations declared on the command line
 	// TODO: add test to ensure that generated annotation has precedence
-	if feedName != nil {
-		annotationArray = append(annotationArray, getFormattedJSON("feed", feedName.GetFullQualifiedName()))
+	if feedQualifiedName != nil {
+		annotationArray = append(annotationArray, getFormattedJSON("feed", feedQualifiedName.GetFullQualifiedName()))
 	}
 	annotations := getParameters(annotationArray, true, true)
 
+	//Within this if statement, we process users' trigger command using the old way.
```
Comment (dubee) [inline commands/trigger.go]: This is quite a large method with over 100 lines of code. Generally large functions are hard to follow and should be broken out into several functions.

File: commands/trigger.go [hunk 11]
```diff
@@ -548,17 +623,30 @@ func (t *Trigger) Update(Client *whisk.Client, args []string) error {
...
 			map[string]interface{}{"name": qualifiedName.GetEntityName(), "err": err})
 		werr := whisk.MakeWskErrorFromWskError(errors.New(errStr), err, whisk.EXIT_CODE_ERR_GENERAL, whisk.DISPLAY_MSG, whisk.NO_DISPLAY_USAGE)
 		return werr
-	} else {
+	}
 
-		// Get full feed name from trigger get request as it is needed to get the feed
-		if retTrigger != nil && retTrigger.Annotations != nil {
-			fullFeedName = getValueString(retTrigger.Annotations, "feed")
+	// Get full feed name from trigger get request as it is needed to get the feed
```
Comment (dubee) [inline commands/trigger.go]: This this update method has 200 lines of code making it extremely hard to understand quickly.

File: commands/trigger.go (hunk 8 above)
Comment (steven0711dong) [reply commands/trigger.go]: OK, I'll have both return error.

File: commands/trigger.go (hunk 10 above)
Comment (steven0711dong) [reply commands/trigger.go]: I can try to separate the 'old way' and 'new way' into two different functions.

**Comment by steven0711dong** [comment]
I've already changed the error message thing, I'll look at how to break them into small functions tonight.

File: tests/src/test/scala/system/basic/WskCliBasicTests.scala [hunk 12]
```diff
@@ -491,6 +491,47 @@ class WskCliBasicTests extends TestHelpers with WskTestHelpers {
...
+      trigger.create(triggerName, parameters = params)
+    }
+    wsk
+      .cli(
+        Seq("trigger", "update", triggerName, "-F", "feedParam", "feedParamVal", "--auth", wskprops.authKey) ++ wskprops.overrides,
+        expectedExitCode = ERROR_EXIT)
+      .stderr should include("this trigger does not contain a feed")
+  }
+
+  it should "return error message when creating or updating feed with both --param and --trigger-param/--feed-param flags" in withAssetCleaner(
```
Comment (dubee) [inline tests/src/test/scala/system/basic/WskCliBasicTests.scala]: Should there also be a test for update?

File: commands/trigger.go [hunk 13]
```diff
@@ -547,15 +640,24 @@ func (t *Trigger) Update(Client *whisk.Client, args []string) error {
 	retTrigger, httpResp, err := Client.Triggers.Get(qualifiedName.GetEntityName())
 
 	if err != nil && httpResp.StatusCode == 404 {
-		t.Create(Client, args)
+		if createErr := t.Create(Client, args); createErr != nil {
```
Comment (dubee) [inline commands/trigger.go]: This is new. If trigger get fails then trigger update now fails, correct?

Comment (steven0711dong) [reply commands/trigger.go]: Correct. 
1. If trigger get fails with 404, then we create the trigger. If creating the triggers fails, we have to report back to the user. Previously we do not report any error when we create trigger in update and that was a bug. 
2. If trigger fails with other reason, execution will fall into next else if, and we report the error message as well.

File: commands/trigger.go [hunk 14]
```diff
@@ -601,3 +703,110 @@ func (t *Trigger) Update(Client *whisk.Client, args []string) error {
...
+//             2. creating a trigger using --param flag
+//then we use the old way to create the trigger.
+func userIndicatesToUseOldTriggerCommand() bool {
+	return len(Flags.common.param) > 0 || (len(Flags.trigger.feedParam) == 0 && len(Flags.common.param) == 0 && len(Flags.trigger.triggerParam) == 0)
+}
+
+func userIssuedNewTriggerCmd() error {
+	if len(Flags.trigger.feedParam) > 0 || len(Flags.trigger.triggerParam) > 0 {
+		whisk.Debug(whisk.DbgError, "User tries to mix use of --param with --feed-param and --trigger-param")
+		errStr := wski18n.T("Incorrect usage. Cannot combine --feed-param or --trigger-param flag with --param flag")
```
Comment (dubee) [inline commands/trigger.go]: Need to say `--param-file` to this list. Is this error message exported to wski18n?

File: commands/trigger.go [hunk 15]
```diff
@@ -397,11 +397,15 @@ func init() {
 	triggerCreateCmd.Flags().StringSliceVarP(&Flags.common.param, "param", "p", []string{}, wski18n.T("parameter values in `KEY VALUE` format"))
 	triggerCreateCmd.Flags().StringVarP(&Flags.common.paramFile, "param-file", "P", "", wski18n.T("`FILE` containing parameter values in JSON format"))
 	triggerCreateCmd.Flags().StringVarP(&Flags.common.feed, "feed", "f", "", wski18n.T("trigger feed `ACTION_NAME`"))
+	triggerCreateCmd.Flags().StringSliceVarP(&Flags.trigger.feedParam, "feed-param", "F", []string{}, wski18n.T("feed parameter values in `KEY VALUE` format"))
```
Comment (dubee) [inline commands/trigger.go]: Are the help entries added to wski18n?

File: tests/src/test/scala/system/basic/WskCliBasicTests.scala (hunk 12 above)
Comment (dubee) [reply tests/src/test/scala/system/basic/WskCliBasicTests.scala]: @steven0711dong , I only see `create` tested here?

File: tests/src/test/scala/system/basic/WskCliBasicTests.scala [hunk 16]
```diff
@@ -491,6 +491,47 @@ class WskCliBasicTests extends TestHelpers with WskTestHelpers {
...
+    wskprops) { (wp, assetHelper) =>
+    val triggerName = withTimestamp("t1tor1")
+    val ns = wsk.namespace.whois()
+
+    var stderr =
```
Comment (dubee) [inline tests/src/test/scala/system/basic/WskCliBasicTests.scala]: Test `-P` as well?

Comment (steven0711dong) [reply tests/src/test/scala/system/basic/WskCliBasicTests.scala]: I added a test for -P

File: commands/trigger.go (hunk 15 above)
Comment (steven0711dong) [reply commands/trigger.go]: added

File: commands/trigger.go (hunk 14 above)
Comment (steven0711dong) [reply commands/trigger.go]: made error msg more generic

File: wski18n/resources/en_US.all.json [hunk 17]
```diff
@@ -1594,5 +1594,21 @@
...
+  {
+    "id": "trigger parameter values in `KEY VALUE` format", 
+    "translation": "trigger parameter values in `KEY VALUE` format"
+  },
+  {
+    "id": "feed parameter values in `KEY VALUE` format", 
+    "translation": "feed parameter values in `KEY VALUE` format"
+  },
+  {
+    "id": "Incorrect usage. trigger without a feed cannot have feed parameters",
```
Comment (dubee) [inline wski18n/resources/en_US.all.json]: Capital `t` in `trigger.

File: commands/trigger.go [hunk 18]
```diff
@@ -431,52 +435,141 @@ func (t *Trigger) Create(Client *whisk.Client, args []string) error {
...
-		annotationArray = append(annotationArray, getFormattedJSON("feed", feedName.GetFullQualifiedName()))
+	if feedQualifiedName != nil {
+		annotationArray = append(annotationArray, getFormattedJSON("feed", feedQualifiedName.GetFullQualifiedName()))
 	}
+
 	annotations := getParameters(annotationArray, true, true)
 
+	//if trigger contains no feed but user tries to update feed parameter, then we issue error.
+	if feedQualifiedName == nil && len(Flags.trigger.feedParam) > 0 {
+		whisk.Debug(whisk.DbgError, "Incorrect usage. trigger without a feed cannot have feed parameters")
```
Comment (dubee) [inline commands/trigger.go]: This error string does not match what is in `en_US.all.json`.
